File: tools/tracer/src/tracer/commands/upstream.py
"""`trace upstream` — what a symbol or path depends on.

Default mode: positional `<symbol>` → transitive dependencies of that symbol
(BFS over forward edges).

Path mode: `--path <path>` → top-N nodes in that path ranked by outgoing
edge count (the highest-coupling code — the symbols that depend on the
most other things).

Both modes return ranked lists from the architecture graph.
"""

from __future__ import annotations

import json
import sys
from collections import Counter
from pathlib import Path

import click

from tracer import architecture
from tracer.deps import require_dependencies


@click.command()
@click.argument("symbol", required=False)
@click.option(
    "--path",
    "path",
    type=click.Path(exists=True),
    default=None,
    help="Switch to path mode: rank top-N most-coupling symbols in this path.",
)
@click.option("--depth", default=3, show_default=True, help="Transitive BFS depth.")
@click.option("--limit", default=10, show_default=True, help="Top-N limit in --path mode.")
@click.option("--json", "as_json", is_flag=True)
def command(
    symbol: str | None,
    path: str | None,
    depth: int,
    limit: int,
    as_json: bool,
) -> None:
    """What a symbol (default) or a path (`--path`) depends on.

    Symbol mode: transitive dependencies of one symbol.
    Path mode: top-N highest-coupling symbols across the path's graph.
    """
    require_dependencies()

    if path is not None:
        _path_mode(path, depth, limit, as_json)
        return

    if not symbol:
        click.echo("Error: pass a SYMBOL or --path <path>.", err=True)
        sys.exit(2)
    _symbol_mode(symbol, depth, as_json)


def _symbol_mode(symbol: str, depth: int, as_json: bool) -> None:
    graph = architecture.get()
    matches = architecture.find_symbols(graph, symbol)
    if not matches:
        click.echo(f"Symbol '{symbol}' not found in architecture graph.", err=True)
        sys.exit(2)

    output: dict[str, dict] = {}
    for match in matches:
        chain = architecture.transitive_dependencies(graph, match.id, max_depth=depth)
        output[match.id] = {
            "symbol": match.label,
            "kind": match.kind,
            "source_file": match.source_file,
            "source_line": match.source_line,
            "dependencies": [
                {
                    "node_id": node.id,
                    "label": node.label,
                    "kind": node.kind,
                    "source_file": node.source_file,
                    "depth": d,
                }
                for node, d in chain
            ],
        }

    if as_json:
        click.echo(json.dumps(output, indent=2))
        return

    for match in matches:
        chain = architecture.transitive_dependencies(graph, match.id, max_depth=depth)
        click.echo(
            f"\n{match.label} [{match.kind}] @ {match.source_file}:{match.source_line}"
        )
        click.echo(f"  depends on (depth ≤ {depth}):")
        if not chain:
            click.echo("    (no architecture-graph dependencies found)")
            continue
        for node, d in chain:
            location = (
                f"{node.source_file}:{node.source_line}"
                if node.source_file
                else "(external)"
            )
            click.echo(f"    [d={d}] {node.label} [{node.kind}] @ {location}")


def _path_mode(path: str, depth: int, limit: int, as_json: bool) -> None:
    """Top-N nodes ranked by outgoing-edge count (highest coupling)."""
    graph = architecture.get(repo_root=Path(path).resolve())
    outgoing = Counter(edge.source for edge in graph.edges)

    ranked = [
        (node_id, count)
        for node_id, count in outgoing.most_common()
        if not node_id.startswith("module::external::")
    ]
    top_ids = [node_id for node_id, _ in ranked[: limit * 3]]

    transitive_counts: dict[str, int] = {}
    for node_id in top_ids:
        transitive_counts[node_id] = len(
            architecture.transitive_dependencies(graph, node_id, max_depth=depth)
        )

    re_ranked = sorted(
        top_ids,
        key=lambda nid: (transitive_counts.get(nid, 0), outgoing.get(nid, 0)),
        reverse=True,
    )[:limit]

    rows = []
    for node_id in re_ranked:
        node = graph.nodes.get(node_id)
        if node is None:
            continue
        rows.append(
            {
                "rank": len(rows) + 1,
                "node_id": node.id,
                "label": node.label,
                "kind": node.kind,
                "source_file": node.source_file,
                "source_line": node.source_line,
                "direct_dependencies": outgoing.get(node_id, 0),
                "transitive_dependencies": transitive_counts.get(node_id, 0),
            }
        )

    if as_json:
        click.echo(
            json.dumps(
                {"path": path, "limit": limit, "depth": depth, "mode": "upstream", "results": rows},
                indent=2,
            )
        )
        return

    if not rows:
        click.echo(
            "(no nodes in the architecture graph — cache may be empty; run `trace cache build`)"
        )
        return

    click.echo(
        f"Top {len(rows)} highest-coupling nodes in {path} (transitive depth ≤ {depth}):"
    )
    click.echo(
        f"  {'#':<3} {'direct':>6}  {'transitive':>10}  {'kind':<10}  symbol @ source"
    )
    for row in rows:
        location = (
            f"{row['source_file']}:{row['source_line']}"
            if row["source_file"]
            else "(no source)"
        )
        click.echo(
            f"  {row['rank']:<3} {row['direct_dependencies']:>6}  "
            f"{row['transitive_dependencies']:>10}  {row['kind']:<10}  "
            f"{row['label']} @ {location}"
        )

File: tools/tracer/src/commands/upstream.rs
//! `trace upstream` — what a symbol or path depends on.
//!
//! Symbol mode: transitive dependencies of one symbol (BFS forward edges).
//! Path mode: top-N highest-coupling symbols across the path's graph.

use crate::{architecture, cache};
use anyhow::Result;
use serde_json::{json, Map, Value};
use std::path::Path;

pub fn run(
    symbol: Option<&str>,
    path: Option<&Path>,
    depth: i64,
    limit: i64,
    as_json: bool,
) -> Result<Value> {
    if let Some(p) = path {
        crate::pathval::require_exists(p, "--path");
        return path_mode(p, depth, limit, as_json);
    }
    let symbol = match symbol {
        Some(s) if !s.is_empty() => s,
        _ => {
            eprintln!("Error: pass a SYMBOL or --path <path>.");
            std::process::exit(2);
        }
    };
    symbol_mode(symbol, depth, as_json)
}

fn symbol_mode(symbol: &str, depth: i64, as_json: bool) -> Result<Value> {
    let here = Path::new(".");
    let repo_root = cache::worktree_root_for(here).unwrap_or_else(|| cache::display_root(here));
    let graph = architecture::get(&repo_root);
    let matches = architecture::find_symbols(&graph, symbol);
    if matches.is_empty() {
        eprintln!("Symbol '{symbol}' not found in architecture graph.");
        std::process::exit(2);
    }

    let mut output = Map::new();
    for m in &matches {
        let chain = architecture::transitive_dependencies(&graph, &m.id, depth);
        let deps: Vec<Value> = chain
            .iter()
            .map(|(node, d)| {
                json!({
                    "node_id": node.id,
                    "label": node.label,
                    "kind": node.kind,
                    "source_file": node.source_file,
                    "depth": d,
                })
            })
            .collect();
        output.insert(
            m.id.clone(),
            json!({
                "symbol": m.label,
                "kind": m.kind,
                "source_file": m.source_file,
                "source_line": m.source_line,
                "dependencies": deps,
            }),
        );
    }

    if !as_json {
        for m in &matches {
            let chain = architecture::transitive_dependencies(&graph, &m.id, depth);
            println!(
                "\n{} [{}] @ {}:{}",
                m.label,
                m.kind,
                m.source_file.as_deref().unwrap_or("None"),
                m.source_line
                    .map(|l| l.to_string())
                    .unwrap_or_else(|| "None".into()),
            );
            println!("  depends on (depth ≤ {depth}):");
            if chain.is_empty() {
                println!("    (no architecture-graph dependencies found)");
                continue;
            }
            for (node, d) in &chain {
                let location = match node.source_file.as_deref() {
                    Some(f) if !f.is_empty() => format!(
                        "{f}:{}",
                        node.source_line
                            .map(|l| l.to_string())
                            .unwrap_or_else(|| "None".into())
                    ),
                    _ => "(external)".to_string(),
                };
                println!("    [d={d}] {} [{}] @ {}", node.label, node.kind, location);
            }
        }
    }
    Ok(Value::Object(output))
}

fn path_mode(path: &Path, depth: i64, limit: i64, as_json: bool) -> Result<Value> {
    // Same logic as downstream::path_mode — the graph must be built
    // against the real git repo root so single-file `--path` arguments
    // don't amputate every cross-file edge.
    let repo_root = cache::worktree_root_for(path).unwrap_or_else(|| cache::display_root(path));
    let graph = architecture::get(&repo_root);

    // Counter(edge.source) — first-seen order, count descending (stable).
    // Path-mode coupling ranks by IMPORT graph only; reference edges are a
    // separate dimension exposed via `references_to` rather than absorbed
    // into module-level fan-out.
    let ranked = ranked_by_edge_count(
        graph
            .edges
            .iter()
            .filter(|e| e.relation == architecture::RELATION_IMPORTS)
            .map(|e| e.source.as_str()),
    );
    let ranked: Vec<(String, usize)> = ranked
        .into_iter()
        .filter(|(id, _)| !id.starts_with("module::external::"))
        .collect();
    let outgoing: std::collections::HashMap<&str, usize> =
        ranked.iter().map(|(id, c)| (id.as_str(), *c)).collect();

    let top_n = (limit * 3).max(0) as usize;
    let top_ids: Vec<String> = ranked
        .iter()
        .take(top_n)
        .map(|(id, _)| id.clone())
        .collect();

    let mut transitive_counts: std::collections::HashMap<String, usize> =
        std::collections::HashMap::new();
    for node_id in &top_ids {
        transitive_counts.insert(
            node_id.clone(),
            architecture::transitive_dependencies(&graph, node_id, depth).len(),
        );
    }

    // Rank by (transitive, outgoing) descending. A stable sort on the
    // negated key tuple keeps the original relative order for equal keys
    // (ties are not reversed — only the comparison is).
    let mut re_ranked = top_ids.clone();
    re_ranked.sort_by(|a, b| {
        let ta = *transitive_counts.get(a).unwrap_or(&0);
        let tb = *transitive_counts.get(b).unwrap_or(&0);
        let oa = *outgoing.get(a.as_str()).unwrap_or(&0);
        let ob = *outgoing.get(b.as_str()).unwrap_or(&0);
        (tb, ob).cmp(&(ta, oa))
    });
    re_ranked.truncate(limit.max(0) as usize);

    let mut rows: Vec<Value> = vec![];
    for node_id in &re_ranked {
        let node = match graph.nodes.get(node_id) {
            Some(n) => n,
            None => continue,
        };
        rows.push(json!({
            "rank": rows.len() + 1,
            "node_id": node.id,
            "label": node.label,
            "kind": node.kind,
            "source_file": node.source_file,
            "source_line": node.source_line,
            "direct_dependencies": *outgoing.get(node_id.as_str()).unwrap_or(&0),
            "transitive_dependencies": *transitive_counts.get(node_id).unwrap_or(&0),
        }));
    }

    let out = json!({
        "path": path.to_string_lossy(),
        "limit": limit,
        "depth": depth,
        "mode": "upstream",
        "results": rows,
    });

    if !as_json {
        if rows.is_empty() {
            println!("(no nodes in the architecture graph — cache may be empty; run `trace cache build`)");
        } else {
            println!(
                "Top {} highest-coupling nodes in {} (transitive depth ≤ {depth}):",
                rows.len(),
                path.to_string_lossy(),
            );
            println!(
                "  {:<3} {:>6}  {:>10}  {:<10}  symbol @ source",
                "#", "direct", "transitive", "kind"
            );
            for row in &rows {
                let location = match row["source_file"].as_str() {
                    Some(f) if !f.is_empty() => format!(
                        "{f}:{}",
                        row["source_line"]
                            .as_i64()
                            .map(|l| l.to_string())
                            .unwrap_or_else(|| "None".into())
                    ),
                    _ => "(no source)".to_string(),
                };
                println!(
                    "  {:<3} {:>6}  {:>10}  {:<10}  {} @ {}",
                    row["rank"].as_i64().unwrap_or(0),
                    row["direct_dependencies"].as_i64().unwrap_or(0),
                    row["transitive_dependencies"].as_i64().unwrap_or(0),
                    row["kind"].as_str().unwrap_or(""),
                    row["label"].as_str().unwrap_or(""),
                    location,
                );
            }
        }
    }
    Ok(out)
}

/// Replicates `collections.Counter(...).most_common()`: count occurrences,
/// then order by count descending with ties in first-seen order.
fn ranked_by_edge_count<'a>(
    items: impl Iterator<Item = &'a str>,
) -> Vec<(String, usize)> {
    let mut counts: std::collections::HashMap<String, usize> =
        std::collections::HashMap::new();
    let mut first_seen: Vec<String> = vec![];
    for it in items {
        let entry = counts.entry(it.to_string()).or_insert_with(|| {
            first_seen.push(it.to_string());
            0
        });
        *entry += 1;
    }
    let mut ranked: Vec<(String, usize)> = first_seen
        .into_iter()
        .map(|k| {
            let c = counts[&k];
            (k, c)
        })
        .collect();
    // Stable sort by count descending; ties keep first-seen order.
    ranked.sort_by(|a, b| b.1.cmp(&a.1));
    ranked
}
